File: client/src/challenge.ts
import { Connection, PublicKey } from "@solana/web3.js";
import { keccak_256 } from "@noble/hashes/sha3";
import { SPANKWALLET_PROGRAM_ID } from "./programId";

// Fase C-opschoning: deze drie helpers stonden eerder gedupliceerd in
// execute.ts, hunt.ts, recovery.ts en initWallet.ts (vier bijna-identieke
// kopieen). Samengevoegd tot dit ene gedeelde bestand - zie STATUS.md.

/**
 * PUNT 2 (STATUS.md sectie 78): regressiebewaking op de vier `new
 * Uint8Array(x)`-kopieerplekken op het Web-Crypto/WebAuthn-tekenpad
 * (initWallet.ts::sha256, webauthnSign.ts::sha256, webauthnSign.ts'
 * `challenge`/`allowCredentials[].id`). Vandaag is elke `x` daar aantoonbaar
 * al een echte Uint8Array (getraceerd tot een `.slice()`/verse-allocatie/
 * hash-output-oorsprong, zie STATUS.md), dus `new Uint8Array(x)` kopieert
 * altijd correct - het gevaar zit uitsluitend in het ANDERE geval (`x` een
 * kale ArrayBuffer, zero-copy over het hele buffer). Deze check maakt dat
 * geen aanname voor de toekomst: als een latere wijziging `x` ooit naar een
 * ArrayBuffer-achtig, byte-verschillend object verandert, faalt dit hard in
 * plaats van stilzwijgend verkeerde bytes te tekenen/hashen.
 */
export function assertByteIdentical(copy: Uint8Array, original: Uint8Array, label: string): void {
  if (copy.length !== original.length) {
    throw new Error(
      `${label}: kopie heeft afwijkende lengte (${copy.length} vs ${original.length}) - PUNT 2-regressie`
    );
  }
  for (let i = 0; i < copy.length; i++) {
    if (copy[i] !== original[i]) {
      throw new Error(`${label}: kopie wijkt af op byte ${i} - PUNT 2-regressie`);
    }
  }
}

export function concatBytes(...arrays: Uint8Array[]): Uint8Array {
  const total = arrays.reduce((sum, a) => sum + a.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const a of arrays) {
    out.set(a, offset);
    offset += a.length;
  }
  return out;
}

export function encodeBorshVecU8(bytes: Uint8Array): Uint8Array {
  const lenBytes = new Uint8Array(4);
  new DataView(lenBytes.buffer).setUint32(0, bytes.length, true);
  return concatBytes(lenBytes, bytes);
}

/**
 * Exacte TS-tegenhanger van build_expected_challenge() in instructions.rs:
 * Keccak-256 (NIET NIST-SHA3) over program_id || wallet || domain || payload.
 */
export function buildExpectedChallenge(
  wallet: PublicKey,
  domain: string,
  payload: Uint8Array
): Uint8Array {
  const domainBytes = new TextEncoder().encode(domain);
  const combined = concatBytes(
    SPANKWALLET_PROGRAM_ID.toBytes(),
    wallet.toBytes(),
    domainBytes,
    payload
  );
  return keccak_256(combined);
}

/**
 * C-1-fix (STATUS.md sectie 69): geeft de 8 LE-bytes terug voor het
 * meebinden van action_nonce vooraan in elke challenge-payload - elke
 * challenge-based instructie (behalve init_wallet zelf, dat al structureel
 * replay-proof is) doet dit als allereerste stap van zijn payload.
 */
export function actionNonceLeBytes(nonce: bigint): Uint8Array {
  const bytes = new Uint8Array(8);
  new DataView(bytes.buffer).setBigUint64(0, nonce, true);
  return bytes;
}

// discriminator(8)+seed_key(33)+wallet_seed_hash(32)+owner_passkey(33)+bump(1)
// +vault_bump(1)+created_at(8)+backup_authority(32) - vaste prefix vóór het
// eerste variabele-lengte veld (recovery_state), zie offsetAfterDepositAuthority.
const OFFSET_RECOVERY_STATE_TAG = 148;
const RECOVERY_STATE_LEN = 41; // initiated_at(8) + new_owner_passkey(33), zonder de 1-byte Option-tag

/**
 * Berekent het offset van WalletAccount.action_nonce - NA twee Option<T>-
 * velden (recovery_state, deposit_authority). Borsh codeert Option::None als
 * exact 1 tagbyte, NIET als de "maximale" ruimte - een vast offset vanaf het
 * begin van het account werkt dus NIET zodra recovery_state/
 * deposit_authority niet allebei Some zijn (het normale geval). Daarom de
 * tag-bytes daadwerkelijk lezen en het echte offset opbouwen, zelfde aanpak
 * als recovery.ts al voor recovery_state deed - empirisch bevestigd nodig
 * tijdens het bouwen van de C-1-fix-tests (STATUS.md sectie 69): een naief
 * vast offset gaf stil een verkeerde (altijd-0) nonce terug.
 *
 * GECORRIGEERD (STATUS.md sectie 129): action_nonce is NIET meer het
 * laatste veld sinds sectie 115/116 - session_epoch, spend_threshold_lamports
 * en disarmed volgen erna. Gedeeld met readSpendThresholdLamports hieronder
 * i.p.v. de tag-walk daar te herhalen.
 */
function offsetAfterDepositAuthority(data: Buffer): number {
  let offset = OFFSET_RECOVERY_STATE_TAG;
  const recoveryStateTag = data[offset];
  offset += 1;
  if (recoveryStateTag === 1) offset += RECOVERY_STATE_LEN;
  offset += 8; // recovery_timelock_seconds: i64, geen Option, altijd aanwezig
  const depositAuthorityTag = data[offset];
  offset += 1;
  if (depositAuthorityTag === 1) offset += 32; // Pubkey
  return offset; // = offset van action_nonce
}

export async function readActionNonce(
  connection: Connection,
  walletPda: PublicKey
): Promise<bigint> {
  const accountInfo = await connection.getAccountInfo(walletPda, "confirmed");
  if (!accountInfo) {
    throw new Error("WalletAccount " + walletPda.toBase58() + " bestaat niet");
  }
  return accountInfo.data.readBigUInt64LE(offsetAfterDepositAuthority(accountInfo.data));
}

/**
 * Leest WalletAccount.spend_threshold_lamports (STATUS.md sectie 127/128/
 * 129, stap A) - action_nonce(8) + session_epoch(8) ná
 * offsetAfterDepositAuthority(), geen Option, dus geen verdere tag-checks
 * nodig. `0n` (de fail-safe default sinds sectie 115) betekent hier
 * expliciet "geen drempel geconfigureerd", zie thresholdBanner.ts.
 */
export async function readSpendThresholdLamports(
  connection: Connection,
  walletPda: PublicKey
): Promise<bigint> {
  const accountInfo = await connection.getAccountInfo(walletPda, "confirmed");
  if (!accountInfo) {
    throw new Error("WalletAccount " + walletPda.toBase58() + " bestaat niet");
  }
  const data = accountInfo.data;
  const offset = offsetAfterDepositAuthority(data) + 8 /* action_nonce */ + 8 /* session_epoch */;
  return data.readBigUInt64LE(offset);
}

File: desktop/src-tauri/src/challenge.rs
// Rust-poort van client/src/challenge.ts - de challenge-berekening verhuist
// hierheen (Tauri-migratie-ontwerp punt 1) zodat een gecompromitteerde
// webview-frontend het niet meer zelf kan bepalen. Gebruikt
// solana_sdk::keccak::hashv - dezelfde functiefamilie als
// programs/spankwallet/src/instructions.rs::build_expected_challenge()
// (regel 538-542) zelf gebruikt, geen losstaande Keccak-implementatie met
// het risico op een verkeerde variant (Keccak vs NIST-SHA3).

use solana_sdk::{keccak, pubkey::Pubkey};

pub const SPANKWALLET_PROGRAM_ID: Pubkey =
    solana_sdk::pubkey!("9ma6vQVA71yUD6jqvyMuYXnMBYGoE7u9bTUbBYEMGBK9");

pub fn build_expected_challenge(wallet: &Pubkey, domain: &[u8], payload: &[u8]) -> [u8; 32] {
    keccak::hashv(&[SPANKWALLET_PROGRAM_ID.as_ref(), wallet.as_ref(), domain, payload]).0
}

pub fn action_nonce_le_bytes(nonce: u64) -> [u8; 8] {
    nonce.to_le_bytes()
}

// WalletAccount-layout (zie client/src/challenge.ts::readActionNonce): vaste
// prefix van 148 bytes (discriminator+seed_key+wallet_seed_hash+
// owner_passkey+bump+vault_bump+created_at+backup_authority), dan
// recovery_state: Option<RecoveryState> (tag + evt. 41 bytes),
// recovery_timelock_seconds: i64 (altijd 8 bytes, geen Option), dan
// deposit_authority: Option<Pubkey> (tag + evt. 32 bytes), dan pas
// action_nonce: u64. Borsh codeert Option::None als exact 1 tagbyte, dus
// een vast offset werkt niet - de tags moeten echt gelezen worden (zelfde
// valkuil als STATUS.md sectie 69 al vaststelde voor de TS-kant).
const OFFSET_RECOVERY_STATE_TAG: usize = 148;
const RECOVERY_STATE_LEN: usize = 41;

pub fn read_action_nonce(data: &[u8]) -> Result<u64, String> {
    let mut offset = OFFSET_RECOVERY_STATE_TAG;
    let recovery_state_tag = *data
        .get(offset)
        .ok_or_else(|| "WalletAccount-data te kort (recovery_state-tag)".to_string())?;
    offset += 1;
    if recovery_state_tag == 1 {
        offset += RECOVERY_STATE_LEN;
    }
    offset += 8; // recovery_timelock_seconds: i64, geen Option

    let deposit_authority_tag = *data
        .get(offset)
        .ok_or_else(|| "WalletAccount-data te kort (deposit_authority-tag)".to_string())?;
    offset += 1;
    if deposit_authority_tag == 1 {
        offset += 32;
    }

    let nonce_bytes: [u8; 8] = data
        .get(offset..offset + 8)
        .ok_or_else(|| "WalletAccount-data te kort (action_nonce)".to_string())?
        .try_into()
        .unwrap();
    Ok(u64::from_le_bytes(nonce_bytes))
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Testvector rechtstreeks gegenereerd met de bestaande, al-in-productie
    /// TS-implementatie (client/src/challenge.ts::buildExpectedChallenge),
    /// via node + @noble/hashes (dezelfde dependency die de browser-client
    /// al gebruikt) - geen aparte Rust-Keccak-implementatie vertrouwen op
    /// blind geloof, maar tegen de daadwerkelijke, al bewezen TS-uitvoer.
    #[test]
    fn matches_typescript_reference_implementation() {
        let wallet = Pubkey::new_from_array([0x11u8; 32]);
        let domain = b"execute";
        let action_nonce: u64 = 7;
        let recipient = Pubkey::new_from_array([0x22u8; 32]);
        let amount_lamports: u64 = 1_000_000_000;

        let mut payload = Vec::new();
        payload.extend_from_slice(&action_nonce_le_bytes(action_nonce));
        payload.extend_from_slice(recipient.as_ref());
        payload.extend_from_slice(&amount_lamports.to_le_bytes());

        let challenge = build_expected_challenge(&wallet, domain, &payload);

        // Gegenereerd met de daadwerkelijke, al-in-productie TS-code
        // (client/src/challenge.ts::buildExpectedChallenge via
        // @noble/hashes/sha3's keccak_256, dezelfde dependency als de
        // browser-client zelf gebruikt) tegen exact dezelfde wallet=0x11*32,
        // domain="execute", payload hieronder.
        let expected_hex = "0becf606d500dc0b3cf6aaf828e01f87240864eb4c035d0763137ba6e86f0270";
        assert_eq!(hex_encode(&challenge), expected_hex);
    }

    fn hex_encode(bytes: &[u8]) -> String {
        bytes.iter().map(|b| format!("{:02x}", b)).collect()
    }

    #[test]
    fn read_action_nonce_handles_all_option_combinations() {
        // recovery_state=None, deposit_authority=None (het gangbare geval)
        let mut data = vec![0u8; 148];
        data.push(0); // recovery_state tag = None
        data.extend_from_slice(&0i64.to_le_bytes()); // recovery_timelock_seconds
        data.push(0); // deposit_authority tag = None
        data.extend_from_slice(&42u64.to_le_bytes()); // action_nonce
        assert_eq!(read_action_nonce(&data).unwrap(), 42);

        // recovery_state=Some, deposit_authority=None
        let mut data2 = vec![0u8; 148];
        data2.push(1); // recovery_state tag = Some
        data2.extend_from_slice(&[0u8; 41]); // recovery_state-inhoud
        data2.extend_from_slice(&0i64.to_le_bytes());
        data2.push(0);
        data2.extend_from_slice(&99u64.to_le_bytes());
        assert_eq!(read_action_nonce(&data2).unwrap(), 99);

        // recovery_state=None, deposit_authority=Some
        let mut data3 = vec![0u8; 148];
        data3.push(0);
        data3.extend_from_slice(&0i64.to_le_bytes());
        data3.push(1); // deposit_authority tag = Some
        data3.extend_from_slice(&[0u8; 32]);
        data3.extend_from_slice(&123u64.to_le_bytes());
        assert_eq!(read_action_nonce(&data3).unwrap(), 123);
    }
}
